plugins/org.eclipse.uml2.uml/src/org/eclipse/uml2/uml/internal/impl/ElementImpl.java

```diff
@@ -189,6 +193,17 @@ public EList<EObject> getStereotypeApplications() {
 		return ElementOperations.getStereotypeApplications(this);
 	}
 
+	public EList<EObject> getStereotypeApplications() {
+		StereotypeApplicationStorage storage = UMLUtil.STEREOTYPE_APPLICATION_STORAGE_IGNORED
```

Comment:
I've spent a few hours investigating ways to optimize this penalty away, but in the end it's generally not possible to reach the outermost package with less than O(n). In theory it would be possible to cache the information about whether a storage is active for an element locally within that element (eFlags), but keeping that cached information up to date would cost a lot in terms of performance and code amount necessary.

I see no way other than following your suggestion to make the new capability (the use of storages) opt-in. I've applied the following changes:

1. In `UMLUtil` I've introduced a `private static boolean stereotypeApplicationStorageEnabled` field, together with a public getter/setter pair.
2. The value of this field defaults to the boolean value of the new system property `org.eclipse.uml2.uml.util.UMLUtil.STEREOTYPE_APPLICATION_STORAGE_ENABLED_DEFAULT`, so the new capability is globally disabled unless this system property is explicitely set to `true` or until `setStereotypeApplicationStorageEnabled(true)` is called.
3. Additionally I've introduced another new system property `org.eclipse.uml2.uml.util.UMLUtil.STEREOTYPE_APPLICATION_STORAGE_ENABLED_OVERRIDE`, which, if set to `true`, globally enables the new capability and makes calls to `setStereotypeApplicationStorageEnabled()` effectless.
4. I've documented the new getter / setter pair.